text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,386 @@
+
+To utilize the command line client, users will be able to install the binary via `cargo install tikv-client`. They will then be able to access the client through the binary `tikv-client`. If they wish for a different name they can alias it in their shell.
+
+## Usage
+
+## Two types of APIs
+TiKV provides two types of APIs for developers:
+- The Raw Key-Value API
+
+    If your application scenario does not need distributed transactions or MVCC (Multi-Version Concurrency Control) and only need to guarantee the atomicity towards one key, you can use the Raw Key-Value API.
+
+- The Transactional Key-Value API
+
+    If your application scenario requires distributed ACID transactions and the atomicity of multiple keys within a transaction, you can use the Transactional Key-Value API.
+
+Generally the Raw Key-Value API has higher throughput and lower latency compare to the Transactional Key-Value API. If distributed ACID transactions is not required, Raw Key-Value API is preferred over Transactional Key-Value API for better performance and ease of use.
+
+The client provides two types of APIs in two separate modules for developers to choose from.
+
+### The common data types
+
+- Key: raw binary data
+- Value: raw binary data
+- KvPair: Key-value pair type
+- KeyRange: Half-open interval of keys
+- Config: Configuration for client
+
+### Try the Raw Key-Value API
+
+To use the Raw Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+```rust
+    let config = Config::new(vec!["127.0.0.1:2379"]);
+```
+
+2. Create a Raw Key-Value client.
+
+```rust
+    let client = RawClient::new(&config);
+```
+
+3. Call the Raw Key-Value client methods to access the data on TiKV. The Raw Key-Value API contains following methods
+
+```rust
+    fn get<K, C>(&self, key: K, cf: C) -> KvFuture<Value>
+    where
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn batch_get<I, K, C>(&self, keys: I, cf: C) -> KvFuture<Vec<KvPair>>
+    where
+        I: IntoIterator<Item = K>,
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn put<P, C>(&self, pair: P, cf: C) -> KvFuture<()>
+    where
+        P: Into<KvPair>,
+        C: Into<Option<String>>;
+
+    fn batch_put<I, P, C>(&self, pairs: I, cf: C) -> KvFuture<()>
+    where
+        I: IntoIterator<Item = P>,
+        P: Into<KvPair>,
+        C: Into<Option<String>>;
+
+    fn delete<K, C>(&self, key: K, cf: C) -> KvFuture<()>
+    where
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn batch_delete<I, K, C>(&self, keys: I, cf: C) -> KvFuture<()>
+    where
+        I: IntoIterator<Item = K>,
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn scan<R, C>(&self, range: R, limit: u32, key_only: bool, cf: C) -> KvFuture<Vec<KvPair>>
```

Comment:
I wonder if it is easier for users if we stick to key/range being the first param and CF always being the second. :thinking:

Reply:
Compare to other parameters, the cf parameter is optional and is it a good idea to have an optional parameter in the middle? Although it is an inappropriate example, but C++ default arguments have to come after normal arguments.

BTW: I have another comment on this signature. Having a KvPair seems to make sense because it might be returned from get. But KeyRange is always input arguments, we can easily break it into two arguments at cost of one more argument. Maybe it's better to just remove this type?

Reply:
It might make sense. :thinking:

Reply:
I was wondering this as well. Maybe @gregwebs has some feedback?